src/Error/WorkerError.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Error;
+
+/**
+ * @author Greg Korba <greg@codito.dev>
+ *
+ * @internal
+ */
+final class WorkerError
```

Comment:
If I understand, those errors are totally _NOT_ related to Error/Error we already have for processing, eg linting failed before fixing, linting failed after linting etc, exception thrown by rule fixer class ?

Those errors (TYPE_INVALID/TYPE_EXCEPTION (during fixing by rule)/TYPE_LINT) were designed to indicate non-blocking part of fixing process (eg if file is not valid and we skip it, nothing "wrong" and we continue) - and overall, each of them are reproducable, if we apply same Fixer version on same codebase, using same PHP ver, result will be same each time.


For those things, eg `RUNNER_ERROR_REPORT` - is it similar, or it's random instability of tcp connection between host and worker? Also, if this issue happen, shall user re-run the command to try again and have the file fixed/checked?
I wonder if we shouldn't break the flow and exit the app if `RUNNER_ERROR_REPORT` happen.

(i focused on `RUNNER_ERROR_REPORT`, but wondering is about any WorkerError)

Reply:
`RUNNER_ERROR_REPORT` is a type of payload send from the worker to the main process, which creates `WorkerError` from decoded data. `WorkerError` is not related to fixing process (the outcome of applying rules to the file), that's why there's dedicated class, other than `Error`. `WorkerError` represents general errors encountered during parallelisation (caught/uncaught exceptions on the worker side + communication problems).

Reply:
I am against naming it "exception" as it's not something meant to throw. It's a simple VO representing problem encountered on worker's side. When it comes to terminating worker's process, I believe it's related to handling `stop-on-violation` in worker command, so if we pass it from the main runner, we can report errors encountered in worker and kill its process, which effectively will lead to ending main process and displaying error summary.

Reply:
stop-on-violation is when we encountered CS issue and we don't want to continue fixing/checking.

WorkerError is saying big issue happen in worker, like TCP problem etc. I do not want to handle such cases in a way that fixing/checking is continuing, but fail the process.
ErrorsManager is not supposed to handle issues outside of touching PHP sources. (fixer::apply and linting around)

Reply:
@keradus OK, so we have 2 scenarios here:

- use current reporting mechanism that reports message, file, line, code and trace (as string), maybe add original class of an error, and instead of collecting these errors in `ErrorsManager` just throw an exception in the main process. This is easy to achieve, but has one drawback: error rendered in the CLI will have always the same exception class (used for throwing, like `WorkerException`), and will have main process' stack trace, not the worker's one (we can use stack trace string sent from worker and append it exception message when verbose mode is enabled, but this won't be as nice as natively rendered stacktrace)
- serialise exception on worker's side, send it through the socket and unserialise in main process. This would allow re-throwing original exception, with original stacktrace which would be better for the user. But there's a problem - serialising throwables is hard, and we need some additional code for that. There is `wyrihaximus/json-throwable`, which currently can't be installed because of `thecodingmachine/safe` requirement (it uses v1, but Infection requires v2), so we could use it when @WyriHaximus adds support for v2.

But before I put an effort into that, I would like to know which way you prefer. Please let me know so we can finish this 🙂.

Reply:
> There is `wyrihaximus/json-throwable`, which currently can't be installed because of `thecodingmachine/safe` requirement (it uses v1, but Infection requires v2), so we could use it when @WyriHaximus adds support for v2.

Support has already been merged, doing some clean up before tagging a new release. Thanks for the nudge to update it 👍 .

Reply:
without seeing code sample it's hard for me to judge @Wirone . But I believe both of your options are about which way to pass exception between 2 processes - please suggest your preference, yet I'm not confident we need whole new dependency to do it, as we know this single particular exception type and can handle it well manually, no need for generic mechanism for that.

(My point is to not pretend to have this exception handled and hide it in ErrorsManager without stopping the flow, no how it's being passed between 2 processes)

Reply:
Error handling was reworked and reviewed, I believe we can resolve this thread now 🙂.